File: Helpers/src/main/java/com/dg/helpers/BigDecimalHelper.java
package com.dg.helpers;

import java.math.BigDecimal;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */
public class BigDecimalHelper
{
    public static BigDecimal withObject(Object value)
    {
        if (value == null) return null;
        else if (value instanceof BigDecimal) return (BigDecimal)value;
        else if (value instanceof Double) return new BigDecimal((Double)value);
        else if (value instanceof Float) return new BigDecimal((Float)value);
        else if (value instanceof Integer) return new BigDecimal((Integer)value);
        else if (value instanceof Short) return new BigDecimal((Short)value);
        else if (value instanceof String) return new BigDecimal((String)value);
        else return null;
    }
}

File: Helpers/src/main/java/com/dg/helpers/BigDecimalHelper.kt
package com.dg.helpers

import java.math.BigDecimal

@Suppress("unused")
object BigDecimalHelper
{
    @Suppress("MemberVisibilityCanBePrivate")
    fun withObject(value: Any?): BigDecimal?
    {
        return if (value == null)
            null
        else value as? BigDecimal ?: when (value)
        {
            is Double -> if (value.isNaN()) null else BigDecimal(value)
            is Float -> if (value.isNaN()) null else BigDecimal(value.toDouble())
            is Int -> BigDecimal(value)
            is Short -> BigDecimal(value.toInt())
            is String -> BigDecimal(value)
            else -> null
        }
    }

    fun withObject(value: Any?, defaultValue: BigDecimal): BigDecimal
    {
        return withObject(value) ?: defaultValue
    }
}
